src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIAssistantClient.cs

```diff
@@ -0,0 +1,379 @@
+            ?.GetValue(assistantClient) as Uri ?? OpenAIChatClient.DefaultOpenAIEndpoint;
+
+        Metadata = new("openai", providerUrl);
+    }
+
+    /// <inheritdoc />
+    public ChatClientMetadata Metadata { get; }
+
+    /// <inheritdoc />
+    public object? GetService(Type serviceType, object? serviceKey = null)
+    {
+        _ = Throw.IfNull(serviceType);
+
+        return
+            serviceKey is not null ? null :
+            serviceType == typeof(AssistantClient) ? _assistantClient :
+            serviceType.IsInstanceOfType(this) ? this :
+            null;
+    }
+
+    /// <inheritdoc />
+    public Task<ChatCompletion> CompleteAsync(
+        IList<ChatMessage> chatMessages, ChatOptions? options = null, CancellationToken cancellationToken = default) =>
+        CompleteStreamingAsync(chatMessages, options, cancellationToken).ToChatCompletionAsync(coalesceContent: true, cancellationToken);
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
+        IList<ChatMessage> chatMessages, ChatOptions? options = null, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        // Extract necessary state from chatMessages and options.
+        (string? assistantId, string? threadId) = GetIds(options);
+        (RunCreationOptions runOptions, List<FunctionResultContent>? toolResults) = CreateRunOptions(chatMessages, options);
+
+        // Ensure we have an assistant, finding one if none was specified, and creating one if none was available.
+        if (assistantId is null)
+        {
+            await foreach (var assistant in _assistantClient.GetAssistantsAsync(cancellationToken: cancellationToken).ConfigureAwait(false))
+            {
+                assistantId = assistant.Id;
+                break;
+            }
+
+            if (assistantId is null)
+            {
+                if (options?.ModelId is string model)
+                {
+                    var assistant = await _assistantClient.CreateAssistantAsync(model, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    assistantId = assistant.Value.Id;
+                }
+                else
+                {
+                    Throw.ArgumentException(nameof(options), "No assistant was available, and no model was supplied for which to create an assistant.");
+                }
+            }
+        }
+
+        // Ensure we have a thread.
+        if (threadId is null)
+        {
+            if (toolResults is not null)
+            {
+                Throw.ArgumentException(nameof(chatMessages), "No thread ID was provided, but chat messages includes tool results.");
+            }
+            else if (options is null)
+            {
+                Throw.ArgumentException(nameof(options), "No thread ID was provided, and no options was supplied to store a new thread's ID into.");
+            }
```

Comment:
Perhaps we could avoid this restriction as per https://github.com/dotnet/extensions/pull/5852/files#r1946310023